compiler/AST/ForallStmt.cpp

```diff
@@ -0,0 +1,204 @@
+  else           INT_ASSERT(fIterVars.length == 1);
+
+  verifyList(fIterVars, this);
+  verifyList(fIterExprs, this);
+  verifyList(fIntentVars, this);
+  for_alist(expr, fIterVars)
+    INT_ASSERT(isDefExpr(expr));
+  for_alist(expr, fIntentVars)
+    INT_ASSERT(isDefExpr(expr));
+
+  INT_ASSERT(fLoopBody);
+  verifyParent(fLoopBody);
+  verifyNotOnList(fLoopBody);
+  // should be a normal block
+  INT_ASSERT(fLoopBody->blockTag == BLOCK_NORMAL);
+  INT_ASSERT(!fLoopBody->blockInfoGet());
+  INT_ASSERT(!fLoopBody->isLoopStmt());
+  INT_ASSERT(!fLoopBody->useList);
+  INT_ASSERT(!fLoopBody->userLabel);
+  INT_ASSERT(!fLoopBody->byrefVars);
+  INT_ASSERT(!fLoopBody->forallIntents);
+
+  INT_ASSERT(fWith);
+
+  // Currently ForallStmt are gone during resolve().
+  INT_ASSERT(!resolved);
+}
+
+void ForallStmt::accept(AstVisitor* visitor) {
+/* TODO once AstVisitor is updated:
+  if (visitor->enterForallStmt(this)) {
+    for_alist(next_ast, fIter)
+      next_ast->accept(visitor);
+    // not doing: fIter->accept(visitor);
+    fWith->acceptFI(visitor); // aka visitor->visitForallIntents(fWith);
+    fMBlock->accept(visitor); // includes loopBody
+    visitor->exitForallStmt(this);
+  }
+*/
+}
+
+Expr* ForallStmt::getFirstChild() {
+  return fIterVars.head;
+}
+
+Expr* ForallStmt::getFirstExpr() {
+  if (/*Expr* iv =*/ fIterVars.head->getFirstExpr())
+    INT_ASSERT(false); // todo: instead return iv;
+  else if (Expr* ie = fIterExprs.head->getFirstExpr())
+    return ie;
+
+  // we must have found something
+  INT_ASSERT(false);
+  return NULL; //dummy
+}
+
+Expr* ForallStmt::getNextExpr(Expr* expr) {
+  if (expr == fIterVars.tail)
+    return fIterExprs.head;
+
+  if (expr == fIterExprs.tail) {
+    if (Expr* inv = fIntentVars.head)
+      return inv;
+    else
+      return fLoopBody->getFirstExpr();
+  }
+
+  // Should we also descend into fWith?
+
+  if (expr == fIntentVars.tail)
+    return fLoopBody->getFirstExpr();
+
+  return this;
+}
+
+// Is 'expr' an iterable-expression for 'this' ?
+bool ForallStmt::isIteratedExpression(Expr* expr) {
+  return expr->list && expr->list == &fIterExprs;
+}
+
+// Return the enclosing forall statement for 'expr', or NULL if none.
+ForallStmt* enclosingForallStmt(Expr* expr) {
+  for (Expr* curr = expr->parentExpr; curr; curr = curr->parentExpr)
+    if (ForallStmt* fs = toForallStmt(curr))
+      return fs;
+  return NULL;
+}
+
+// If 'var' is listed in the this's with-clause with a reduce intent,
+// return its position in the with-clause, otherwise return -1.
+int ForallStmt::reduceIntentIdx(Symbol* var) {
+  ForallIntents* fi = withClause();
+  int nv = fi->numVars();
+  for (int i = 0; i < nv; i++)
+    if (fi->isReduce(i))
+      if (SymExpr* varSE = toSymExpr(fi->fiVars[i]))
+        if (varSE->symbol() == var)
+          return i;
+
+  // Did not see 'var' with a reduce intent.
+  return -1;
+}
+
+GenRet ForallStmt::codegen() {
+  INT_ASSERT(false); // should not be invoked
+  return (GenRet)0;
```

Comment:
I'm not so sure what (GenRet)0 actually does. I'd normally write
```
GenRet ret;
return ret;
```

but that is a nit since the compiler should quit before getting there anyway.